internal/service/gateway/connection.go

```diff
@@ -69,9 +75,88 @@ func ResourceConnection() *schema.Resource {
 				},
 			},
 		},
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Type:    resourceConnectionResourceV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: resourceConnectionStateUpgradeV0,
+				Version: 0,
+			},
+		},
 	}
 }
 
+func resourceConnectionResourceV0() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description:      "The name of the connection, should be unique within the gateway.",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validateName,
+			},
+			"gateway": {
+				Description: "The ID of the upcloud_gateway resource to which the connection belongs.",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"type": {
+				Description: "The type of the connection; currently the only supported type is 'ipsec'.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Default:     "ipsec",
+				ValidateDiagFunc: validation.ToDiagFunc(
+					validation.StringInSlice([]string{"ipsec"}, false),
+				),
+			},
+			"local_route": {
+				Description:  "Route for the UpCloud side of the network.",
+				Type:         schema.TypeSet,
+				Optional:     true,
+				AtLeastOneOf: []string{"local_route", "remote_route"},
+				Elem:         gatewayRouteSchema(),
+			},
+			"remote_route": {
+				Description:  "Route for the remote side of the network.",
+				Type:         schema.TypeSet,
+				Optional:     true,
+				AtLeastOneOf: []string{"local_route", "remote_route"},
+				Elem:         gatewayRouteSchema(),
+			},
+			"tunnels": {
+				Description: "List of connection's tunnels names. Note that this field can have outdated information as connections are created by a separate resource. To make sure that you have the most recent data run 'terrafrom refresh'.",
```

Comment:
Does this now take in UUIDs
```suggestion
				Description: "List of connection's tunnels UUIDs. Note that this field can have outdated information as connections are created by a separate resource. To make sure that you have the most recent data run 'terrafrom refresh'.",
```

Reply:
Still names. We could revise this in further updates.